packages/flutter/lib/src/material/bottom_navigation_bar.dart

```diff
@@ -250,23 +269,48 @@ class BottomNavigationBar extends StatefulWidget {
   /// If null then the [TextTheme.caption]'s color is used.
   final Color unselectedItemColor;
 
-  /// The font size of the [BottomNavigationBarItem] labels when they are selected.
+  /// The size, opacity, and color of the icon in the currently selected
+  /// [BottomNavigationBarItem.icon].
   ///
-  /// Defaults to `14.0`.
-  final double selectedFontSize;
+  /// If this is not provided, the size will default to [iconSize], the color
+  /// will default to [selectedItemColor].
+  ///
+  /// It this field is provided, it must contain non-null [IconThemeData.size]
+  /// and [IconThemeData.color] properties. Also, if this field is supplied,
+  /// [unselectedIconTheme] must be provided.
+  final IconThemeData selectedIconTheme;
 
-  /// The font size of the [BottomNavigationBarItem] labels when they are not
-  /// selected.
+  /// The size, opacity, and color of the icon in the currently unselected
+  /// [BottomNavigationBarItem.icon]s
   ///
-  /// Defaults to `12.0`.
-  final double unselectedFontSize;
+  /// If this is not provided, the size will default to [iconSize], the color
+  /// will default to [unselectedItemColor].
+  ///
+  /// It this field is provided, it must contain non-null [IconThemeData.size]
+  /// and [IconThemeData.color] properties. Also, if this field is supplied,
+  /// [unselectedIconTheme] must be provided.
+  final IconThemeData unselectedIconTheme;
+
+  /// The [TextStyle] of the [BottomNavigationBarItem] labels when they are
+  /// selected.
+  final TextStyle selectedLabelStyle;
+
+  /// The [TextStyle] of the [BottomNavigationBarItem] labels when they are not
+  /// selected.
+  final TextStyle unselectedLabelStyle;
 
   /// Whether the labels are shown for the selected [BottomNavigationBarItem].
   final bool showUnselectedLabels;
 
   /// Whether the labels are shown for the unselected [BottomNavigationBarItem]s.
   final bool showSelectedLabels;
 
+  /// Convenience getter for knowing the font size of the selected [BottomNavigationBarItem].
+  double get selectedFontSize => selectedLabelStyle.fontSize;
```

Comment:
It would be more conventional, per related comments, to leave these out and restore the original [un]selectedFontSize properties.

Reply:
I mentioned this above, but it feels a little weird to me to have both `selectedFontSize` and `selectedTextStyle` as properties. Since I have to merge `selectedFontSize` and `selectedTextStyle` at some point, I think it makes the most sense to do it once (in initializer list of the constructor) and then keep `selectedTextStyle` as the SSOT.

Reply:
Another way to view the widget class is just as a snapshot of the widget's configuration, i.e. its constructor parameters. So the truth is exactly the parameters the widget was configured with.

Reply:
OK, I still slightly prefer the constructor technique, but your way does seem more consistent. So I moved the `_effectiveTextStyle` logic to the `_BottomNavigationTile` `build` method, and I left `BottomNavigationBar` as a simple holder of values.